documentation/0029-node-event-loop/index.md

```diff
@@ -0,0 +1,185 @@
+
+![Call stack first example](call-stack-first-example.png)
+
+The event loop on every iteration looks if there's something in the call stack, and executes it:
+
+![Execution order first example](execution-order-first-example.png)
+
+until the call stack is empty.
+
+## Queuing function execution
+
+The above example looks normal, there's nothing special about it: JavaScript finds things to execute, runs them in order.
+
+Let's see how to defer a function until the stack is clear.
+
+The use case of `setTimeout(() => {}), 0)` is to call a function, but execute it once every other function in the code has executed.
+
+Take this example:
+
+```js
+const bar = () => console.log('bar')
+
+const baz = () => console.log('baz')
+
+const foo = () => {
+  console.log('foo')
+  setTimeout(bar, 0)
+  baz()
+}
+
+foo()
+```
+
+This code prints, maybe surprisingly:
+
+```txt
+foo
+baz
+bar
+```
+
+When this code runs, first foo() is called. Inside foo() we first call setTimeout, passing `bar` as an argument, and we instruct it to run immediately as fast as it can, passing 0 as the timer. Then we call baz().
+
+At this point the call stack looks like this:
+
+![Call stack second example](call-stack-second-example.png)
+
+Here is the execution order for all the functions in our program:
+
+![Execution order second example](execution-order-second-example.png)
+
+Why is this happening?
+
+## The Message Queue
+
+When setTimeout() is called, the Browser or Node.js start the timer. Once the timer expires, in this case immediately as we put 0 as the timeout, the callback function is put in the **Message Queue**.
+
+The Message Queue is also where user-initiated events like click or keyboard events, or fetch responses are queued before your code has the opportunity to react to them. Or also DOM events like `onLoad`.
+
+**The loop gives priority to the call stack, and it first processes everything it finds in the call stack, and once there's nothing in there, it goes to pick up things in the event queue.**
+
+We don't have to wait for functions like `setTimeout`, fetch or other things to do their own work, because they are provided by the browser, and they live on their own threads. For example, if you set the `setTimeout` timeout to 2 seconds, you don't have to wait 2 seconds - the wait happens elsewhere.
+
+## ES6 Job Queue
+
+ECMAScript 2015 introduced the concept of the Job Queue, which is used by Promises (also introduced in ES6/ES2015). It's a way to execute the result of an async function as soon as possible, rather than being put at the end of the call stack.
+
+Promises that resolve before the current function ends will be executed right after the current function.
+
+I find nice the analogy of a rollercoaster ride at an amusement park: the message queue puts you back in queue with after all the other people in the queue, while the job queue is the fastpass ticket that lets you take another ride right after you finished the previous one.
```

Comment:
'with after' seems like a grammatical error. This can be written as ` The message queue puts you at the back of the queue, behind all the other people, where you will have to wait for your turn, while the job queue is the fastpass ticket that lets you take another ride right after you finished the previous one.`

Reply:
👍